src/mds/objects.py

```diff
@@ -0,0 +1,236 @@
+from collections import Iterable
+from enum import Enum
+
+from authutils.token.fastapi import access_token
+from asyncpg import UniqueViolationError
+from fastapi import Depends, HTTPException, APIRouter, Security
+from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
+from gen3authz.client.arborist.async_client import ArboristClient
+import httpx
+from starlette.requests import Request
+from starlette.responses import JSONResponse
+from starlette.status import (
+    HTTP_201_CREATED,
+    HTTP_409_CONFLICT,
+    HTTP_400_BAD_REQUEST,
+    HTTP_401_UNAUTHORIZED,
+    HTTP_403_FORBIDDEN,
+    HTTP_500_INTERNAL_SERVER_ERROR,
+)
+from pydantic import BaseModel
+
+from . import config, logger
+from .models import Metadata
+
+mod = APIRouter()
+arborist = ArboristClient()
+
+# auto_error=False prevents FastAPI from raises a 403 when the request is missing
+# an Authorization header. Instead, we want to return a 401 to signify that we did
+# not recieve valid credentials
+bearer = HTTPBearer(auto_error=False)
+
+
+class FileUploadStatus(str, Enum):
+    NOT_STARTED = "not_uploaded"
+    DONE = "uploaded"
+    ERROR = "error"
+
+
+class CreateObjInput(BaseModel):
+    """
+    Create object.
+
+    file_name (str): Name for the file being uploaded
+    authz (dict): authorization block with requirements for what's being uploaded
+    aliases (list, optional): unique name to allow using in place of whatever GUID gets
+        created for this upload
+    metadata (dict, optional): any additional metadata to attach to the upload
+    """
+
+    file_name: str
+    authz: dict
+    aliases: list = None
+    metadata: dict = None
+
+
+@mod.post("/objects")
+async def create_object(
+    body: CreateObjInput,
+    request: Request,
+    token: HTTPAuthorizationCredentials = Security(bearer),
+):
+    """
+    Create object placeholder and attach metadata, return Upload url to the user.
+
+    Args:
+        body (CreateObjInput): input body for create object
+    """
+    try:
+        # NOTE: token can be None if no Authorization header was provided, we expect
+        #       this to cause a downstream exception since it is invalid
+        token_claims = await access_token("user", "openid", purpose="access")(token)
+    except Exception as exc:
+        logger.error(exc, exc_info=True)
+        raise HTTPException(
+            HTTP_401_UNAUTHORIZED,
+            f"Could not verify, parse, and/or validate scope from provided access token.",
+        )
+
+    file_name = body.dict().get("file_name")
+    authz = body.dict().get("authz")
+    aliases = body.dict().get("aliases") or []
+    metadata = body.dict().get("metadata")
+    logger.debug(f"validating authz block input: {authz}")
```

Comment:
(not a big deal) https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/